File: frontend/src/app/services/student-analytics.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, of } from 'rxjs';
import { map, catchError } from 'rxjs/operators';
import { environment } from '../../environments/environment';
import { GamificationService } from './gamification.service';
import { QuizService } from '../core/services/quiz.service';

export interface StudentAnalyticsData {
  // Données pour le ML
  previousAttempts: number;
  credits: number;
  totalClicks: number;
  sessions: number;
  avgClicks: number;
  maxClicks: number;
  avgScore: number;
  minScore: number;
  maxScore: number;
  assessments: number;
  registrationDate: number;
  isUnregistered: number;
  
  // Métadonnées
  dataQuality: 'high' | 'medium' | 'low';
  lastUpdated: Date;
}

@Injectable({
  providedIn: 'root'
})
export class StudentAnalyticsService {
  private apiUrl = environment.apiUrl;

  constructor(
    private http: HttpClient,
    private gamificationService: GamificationService,
    private quizService: QuizService
  ) {}

  /**
   * Récupère et agrège toutes les données d'un étudiant pour le ML
   * Utilise maintenant la nouvelle API StudentAnalytics pour les vraies données
   */
  getStudentAnalytics(studentId: number): Observable<StudentAnalyticsData> {
    // Essayer d'abord de récupérer les vraies analytics depuis la BD
    return this.http.get<any>(`${this.apiUrl}/analytics/student/${studentId}`).pipe(
      map(analytics => this.convertToAnalyticsData(analytics)),
      catchError(error => {
        console.warn('Analytics API non disponible', error);
        // Retourner des données par défaut
        return of(this.getDefaultAnalytics());
      })
    );
  }

  /**
   * Convertit les analytics de la BD vers le format ML
   */
  private convertToAnalyticsData(analytics: any): StudentAnalyticsData {
    // Calculer les jours depuis l'inscription
    const registrationDate = analytics.firstRegistrationDate
      ? -Math.floor((Date.now() - new Date(analytics.firstRegistrationDate).getTime()) / (1000 * 60 * 60 * 24))
      : -30;

    return {
      previousAttempts: analytics.previousAttempts || 0,
      credits: analytics.studiedCredits || 0,
      totalClicks: analytics.totalClicks || 0,
      sessions: analytics.totalSessions || 0,
      avgClicks: analytics.avgClicksPerSession || 0,
      maxClicks: analytics.maxClicksInSession || 0,
      avgScore: analytics.avgScore || 0,
      minScore: analytics.minScore || 0,
      maxScore: analytics.maxScore || 0,
      assessments: analytics.totalAssessments || 0,
      registrationDate: registrationDate,
      isUnregistered: analytics.isUnregistered ? 1 : 0,
      dataQuality: this.assessRealDataQuality(analytics),
      lastUpdated: new Date()
    };
  }

  /**
   * Évalue la qualité des vraies données
   */
  private assessRealDataQuality(analytics: any): 'high' | 'medium' | 'low' {
    const hasClicks = analytics.totalClicks > 0;
    const hasSessions = analytics.totalSessions > 0;
    const hasAssessments = analytics.totalAssessments > 0;
    const hasScores = analytics.avgScore > 0;

    const score = [hasClicks, hasSessions, hasAssessments, hasScores].filter(Boolean).length;

    if (score >= 3) return 'high';
    if (score >= 2) return 'medium';
    return 'low';
  }

  /**
   * Retourne des analytics par défaut
   */
  private getDefaultAnalytics(): StudentAnalyticsData {
    return {
      previousAttempts: 0,
      credits: 0,
      totalClicks: 0,
      sessions: 0,
      avgClicks: 0,
      maxClicks: 0,
      avgScore: 0,
      minScore: 0,
      maxScore: 0,
      assessments: 0,
      registrationDate: -30,
      isUnregistered: 0,
      dataQuality: 'low',
      lastUpdated: new Date()
    };
  }

  // ========== MÉTHODES DE TRACKING EN TEMPS RÉEL ==========

  /**
   * Track un clic de l'étudiant
   */
  trackClick(userId: number, clicks: number = 1): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/analytics/student/${userId}/click`, { clicks })
      .pipe(
        catchError(error => {
          console.error('Erreur lors du tracking du clic:', error);
          return of(void 0);
        })
      );
  }

  /**
   * Track une session de l'étudiant (à appeler au login)
   */
  trackSession(userId: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/analytics/student/${userId}/session`, {})
      .pipe(
        catchError(error => {
          console.error('Erreur lors du tracking de la session:', error);
          return of(void 0);
        })
      );
  }

  /**
   * Track une évaluation complétée
   */
  trackAssessment(userId: number, score: number, type: 'TMA' | 'CMA' | 'EXAM'): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/analytics/student/${userId}/assessment`, { score, type })
      .pipe(
        catchError(error => {
          console.error('Erreur lors du tracking de l\'évaluation:', error);
          return of(void 0);
        })
      );
  }

  /**
   * Ajoute des crédits étudiés (à appeler quand un cours est complété)
   */
  addStudiedCredits(userId: number, credits: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/analytics/student/${userId}/credits`, { credits })
      .pipe(
        catchError(error => {
          console.error('Erreur lors de l\'ajout des crédits:', error);
          return of(void 0);
        })
      );
  }

  /**
   * Incrémente les tentatives (à appeler quand un étudiant échoue un quiz)
   */
  incrementAttempts(userId: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/analytics/student/${userId}/attempt`, {})
      .pipe(
        catchError(error => {
          console.error('Erreur lors de l\'incrémentation des tentatives:', error);
          return of(void 0);
        })
      );
  }

  /**
   * Track l'ouverture d'une leçon
   */
  trackLessonOpened(userId: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/analytics/student/${userId}/lesson-opened`, {})
      .pipe(
        catchError(error => {
          console.error('Error tracking lesson opened:', error);
          return of(void 0);
        })
      );
  }

  /**
   * Ajoute du temps passé sur une leçon (en minutes)
   */
  addTimeSpent(userId: number, minutes: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/analytics/student/${userId}/time-spent`, { minutes })
      .pipe(
        catchError(error => {
          console.error('Error adding time spent:', error);
          return of(void 0);
        })
      );
  }
}

File: backend/auth-service/src/main/java/com/englishflow/auth/service/StudentAnalyticsService.java
package com.englishflow.auth.service;

import com.englishflow.auth.entity.StudentAnalytics;
import com.englishflow.auth.repository.StudentAnalyticsRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.Optional;

@Service
@RequiredArgsConstructor
@Slf4j
public class StudentAnalyticsService {

    private final StudentAnalyticsRepository analyticsRepository;

    /**
     * Récupère ou crée les analytics d'un étudiant
     */
    @Transactional
    public StudentAnalytics getOrCreateAnalytics(Long userId) {
        return analyticsRepository.findByUserId(userId)
                .orElseGet(() -> {
                    log.info("Création des analytics pour l'étudiant {}", userId);
                    StudentAnalytics analytics = StudentAnalytics.builder()
                            .userId(userId)
                            .firstRegistrationDate(LocalDateTime.now())
                            .build();
                    return analyticsRepository.save(analytics);
                });
    }

    /**
     * Enregistre un clic de l'étudiant
     */
    @Transactional
    public void trackClick(Long userId, int clicks) {
        StudentAnalytics analytics = getOrCreateAnalytics(userId);
        analytics.incrementClicks(clicks);
        analytics.updateMaxClicks(clicks);
        analytics.recalculateAvgClicks(); // Recalculer après ajout des clics
        analyticsRepository.save(analytics);
        log.debug("Clics trackés pour l'étudiant {}: {} clics", userId, clicks);
    }

    /**
     * Enregistre une nouvelle session
     */
    @Transactional
    public void trackSession(Long userId) {
        StudentAnalytics analytics = getOrCreateAnalytics(userId);
        analytics.incrementSession();
        analytics.recalculateAvgClicks(); // Recalculer après ajout de session
        analyticsRepository.save(analytics);
        log.debug("Session trackée pour l'étudiant {}", userId);
    }

    /**
     * Enregistre un score d'évaluation
     */
    @Transactional
    public void trackAssessment(Long userId, double score, String assessmentType) {
        StudentAnalytics analytics = getOrCreateAnalytics(userId);
        analytics.addAssessmentScore(score, assessmentType);
        analyticsRepository.save(analytics);
        log.info("Évaluation trackée pour l'étudiant {}: {} - score: {}", 
                userId, assessmentType, score);
    }

    /**
     * Ajoute des crédits étudiés
     */
    @Transactional
    public void addStudiedCredits(Long userId, int credits) {
        StudentAnalytics analytics = getOrCreateAnalytics(userId);
        analytics.addCredits(credits);
        analyticsRepository.save(analytics);
        log.info("Crédits ajoutés pour l'étudiant {}: {} crédits", userId, credits);
    }

    /**
     * Incrémente le nombre de tentatives
     */
    @Transactional
    public void incrementAttempts(Long userId) {
        StudentAnalytics analytics = getOrCreateAnalytics(userId);
        analytics.incrementAttempts();
        analyticsRepository.save(analytics);
        log.info("Tentative incrémentée pour l'étudiant {}", userId);
    }

    /**
     * Marque un étudiant comme désinscrit
     */
    @Transactional
    public void markAsUnregistered(Long userId) {
        StudentAnalytics analytics = getOrCreateAnalytics(userId);
        analytics.setIsUnregistered(true);
        analyticsRepository.save(analytics);
        log.info("Étudiant {} marqué comme désinscrit", userId);
    }

    /**
     * Track l'ouverture d'une leçon
     */
    @Transactional
    public void trackLessonOpened(Long userId) {
        StudentAnalytics analytics = getOrCreateAnalytics(userId);
        analytics.trackLessonOpened();
        analyticsRepository.save(analytics);
        log.debug("Leçon ouverte trackée pour l'étudiant {}", userId);
    }

    /**
     * Ajoute du temps passé sur une leçon
     */
    @Transactional
    public void addTimeSpent(Long userId, int minutes) {
        StudentAnalytics analytics = getOrCreateAnalytics(userId);
        analytics.addTimeSpent(minutes);
        analyticsRepository.save(analytics);
        log.debug("Temps ajouté pour l'étudiant {}: {} minutes", userId, minutes);
    }

    /**
     * Récupère les analytics d'un étudiant
     */
    public Optional<StudentAnalytics> getAnalytics(Long userId) {
        return analyticsRepository.findByUserId(userId);
    }

    /**
     * Supprime les analytics d'un étudiant
     */
    @Transactional
    public void deleteAnalytics(Long userId) {
        analyticsRepository.deleteByUserId(userId);
        log.info("Analytics supprimées pour l'étudiant {}", userId);
    }
}
